design/expected-hardware-configuration-validation.md

```diff
@@ -0,0 +1,209 @@
+           type: Object
+           disk:
+               properties:
+                   sizeBytesGB:
+                       type: Integer
+                   numberOfDisks:
+                       type: Integer
+           type: Object
+           nics:
+               properties:
+                  numberofNICS:
+                     type: Integer
+           type: Object
+           ram:
+               properties:
+                   sizeBytesGB:
+                       type: Integer
+           type: Object
+           systemVendor:
+              properties:
+                  name:
+                      type: String
+           type: Object
+           firmware:
+              properties:
+                  version:
+                      RAID:
+                          type: String
+                      BasebandManagement:
+                          type: String
+                      BIOS: 
+                          type: String
+                      IDRAC: 
+                          type: String
+            type: Object
+        required:
+        - filter
+        - namespace
+    ```
+* Write a new API as hardware-classification-controller/api/v1alpha1 and new kind(CRD) HardwareClassificationController.
+    e.g.
+        
+       kubebuilder create api --group metal3.io --version v1alpha1 --kind HardwareClassificationController
+    
+    - This will create the files api/v1alpha1/hardwareClassificationController_types.go where the API is defined and the controller/hardwareClassificationController_controller.go where the reconciliation business logic is implemented for this Kind(CRD).
+
+    - Implement a new function fetchHost() which will fetch all baremetal hosts from Baremetal-Operator of namespace 'metal3' provided by user in CRD_yaml. This function will return a list of filtered hosts(hosts in status 'ready' or 'inspecting'). 
+
+    - In hardwareClassificationController_controller.go, reconcile function will call fetchHost()        function to fetch all baremetal hosts and also extract hardware configuration, namespace and filter from `metal3.io_HardwareClassificationController_crd.yaml`.
+
+    - Create a classification_manager.go file. Create a function manager() which will have inputs namespace, filter and BMH host_list. This function will call appropriate comparision logic written in classification_filter.go based on 'filter' specified by user in CRD_yaml.
+
+    - The classification_filter.go file will contain multiple comparision algorithms based on filters(i.e minimum and maximum). Create a ClassificationComparison() function which will have inputs fetched baremetalhost list and extracted hardware profile. Comparision function will check each host against profile and update the label if match found.
+
+
+* Create the Schema struct for `ExpectedHardwareConfiguration` inside `HardwareClassificationControllerSpec`,
+in file /api/v1alpha1/hardwareClassificationController_types.go.
+
+    ```yaml
```

Comment:
I still think this is too limited. I would like for us to explore options for using more dynamic expressions in the CRD, so we don't have to add fields to it later in order to support other types of comparisons or tests. Here are a few expression parsing libraries that look interesting:

- https://github.com/antonmedv/expr 
- https://github.com/google/cel-go
- https://github.com/PaesslerAG/gval

The first one looks like it has support for reflection-based expressions.

We could also build something based on JSONPath syntax, since that is common elsewhere in Kubernetes, but it would mean building our own code to evaluate the operators and that might be more work than adopting one of the syntaxes supported by the existing libraries.

Reply:
Okay. I have gone through the expression concept, it looks interesting technique. I would like to keep customFilters also to implement more specific scenarios. I will go with https://github.com/antonmedv/expr for implementing expression based filter.
Here is a thing - 
We can add both filters in CR  - 
ClassificationFilters - 
   - CustomFilters: MinimumHardware
   - Expression: 'RAM >= 20' or  'RAM == 50' or 'RAM <= 96'

Reply:
Before you do a bunch of work with that expression library, the feedback I've had from some of our internal experts is that we don't want to build a DSL as part of the API. It's better to do something like

```
- rules:
  - field: someNameHere
    operation: Equals
    value: 5
```

They gave as reference the way resources are exposed on a node https://github.com/kubernetes/api/blob/master/core/v1/types.go#L4440-L4447 or "extended resources" on a node https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#extended-resources, which are used for things like device plugins https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/

Reply:
We will update the proposal based on above inputs.

Reply:
@dhellmann  We have updated proposal by adding two option for user to use anyof classification controller or expression rules filters. Also added representional and tasks in the proposal.